fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMerger.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * IndexingMerger.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.FDBException;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
+ * Implement an index merge.
+ * This can be either used for an explicit index merge call or during online indexing. In the second case,
+ * keeping throttle values between calls has an overall impact on performance.
+ */
+@ParametersAreNonnullByDefault
+public class IndexingMerger {
+    private static final Logger LOGGER = LoggerFactory.getLogger(IndexingMerger.class);
+    private final Index index;
+    private long mergesLimit;
+    private int mergeSuccesses = 0;
+    private final IndexingCommon common;
+
+    public IndexingMerger(final Index index,  IndexingCommon common, long initialMergesCountLimit) {
+        this.index = index;
+        this.common = common;
+        this.mergesLimit = initialMergesCountLimit;
+    }
+
+    private CompletableFuture<FDBRecordStore> openRecordStore(@Nonnull FDBRecordContext context) {
+        return common.getRecordStoreBuilder().copyBuilder().setContext(context).openAsync();
+    }
+
+    @SuppressWarnings("squid:S3776") // cognitive complexity is high, candidate for refactoring
+    CompletableFuture<Void> mergeIndex() {
+        final AtomicInteger failureCountLimit = new AtomicInteger(1000);
+        AtomicReference<IndexDeferredMaintenanceControl> mergeControlRef = new AtomicReference<>();
+        return AsyncUtil.whileTrue(() ->
```

Comment:
✅ 
Note that: 
1. The event will be defined in `FDBStoreTimer.Events` to avoid a complicated dependency. 
2. To keep it general (as the merger knows not about Lucene), I’ve used the event name `MERGE_INDEX`
Is that acceptable?

Reply:
I was thinking it should be around the entire future, including retries, but this will probably work fine.